### PR TITLE
XTR training refactor

### PR DESCRIPTION
**Pull request by NohTow**
This PR builds upon #207, enhancing it to be better integrated into PyLate.
Essentially, the goal of the PR is to add XTR into PyLate, both training and retrieval (as a two-step retrieval mecanism, using either Voyager or SCANN, the "PLAID-version" of retrieval will be added in the form of WARP in #211).

The PR is a bit heavy but a lot of changes are tests (mostly Claude built, ngl) and I'll try to describe all the other changes.
- The first part is adding XTR scoring (for training, as they differ as the ones used for retrieval)). They are pretty similar to colbert scoring but with the additional masking. As they require some parameters, they are now defined as class rather than pure callable (but still define __call__, so still a callable). I believe we could maybe better re-use part of ColBERT into those (call ColBERT then mask), but it can be done later on.
- One of the biggest change is due to the fact that, unlike ColBERT, XTR *requires* all documents at training time to perform scoring (top-k filtering). To have an unified loss definition between ColBERT/XTR rather than two branches, I refactored the scoring part of contrastive loss to iterate over queries while feeding all documents at the scoring time.  We can still chunk the documents at scoring time (to avoid transient memory spike because of the big einsum) by defining document_chunk_size in the scoring function when defining the loss. We do not feed the whole set of queries at once to be able release some memory in GradCache when doing the backward pass. This is quite important, although we could also give the possibility to the user to use gradient checkpointing to avoid iterating at the cost of one additional pass (in addition to the gradcache one as well :/). This chunking dimension (over the query) is definable by the `scoring_micro_batch_size` value, so we can define the scoring chunking granularly. It also allowed to extends the behavior to the standard contrastive, which is cool because memory used during scoring was quite heavy (though I did not add document_chunk_size and kept the old per row iteration, we can change this later).

As discussed in DMs, I ran a bunch of tests and the training looks very healthy.
- Modified Voyager `__call__` output format to match the one of SCANN. There was a disrepency and the SCANN one was actually a better format because Voyager supposed all queries had the same lengths, which is less and less true with new models and would either crash or silently propagate mistakes. Only downside is if someone was using the output as a numpy array, but 1. they should not use it directly and 2. we are deprecating Voyager anyways.
- It adds XTR retrieve and associated scoring mechanism. Here I am still not satisfied entirely. I made the check over single stage and multiple stage indexes less brittle by adding a dedicated property, but I still think it is kind of brittle and will create odd things such as calling XTR retrieve on PLAID indexes returning ColBERT results. I think we can rather try to define a property saying what "algorithm" we are using (None for two stage, PLAID, XTR) and match it between index/retrieve function, but I am pretty sure we can do an even better integration. We can still merge somehow like this and deprecate XTR.retrieve later by just calling the new mechanism when calling this.

### CODE REVIEW

File: pylate/retrieve/xtr.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    ...     device="cpu",
+    ... )
+
+    >>> assert isinstance(results, list)
+    >>> assert len(results) == 1
+
+    """
+
+    def __init__(self, index: indexes.Base, verbose: bool = False) -> None:
+        if isinstance(index, indexes.PLAID):
```
Comment (NohTow) [inline pylate/retrieve/xtr.py]: it'll break when we add WARP no?

File: pylate/indexes/__init__.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
 from __future__ import annotations
 
+from .base import Base
 from .plaid import PLAID
 from .scann import ScaNN
 from .voyager import Voyager
 
-__all__ = ["Voyager", "PLAID", "ScaNN"]
+__all__ = ["Base", "Voyager", "PLAID", "ScaNN"]
```
Comment (raphaelsty) [inline pylate/indexes/__init__.py]: Why do we want to expose Base ? Unless a good reason, better not to expose it

File: pylate/indexes/plaid.py [hunk 3]
```diff
@@ -115,6 +115,8 @@ class PLAID(Base):
 
     """
 
+    is_end_to_end_index = True
```
Comment (raphaelsty) [inline pylate/indexes/plaid.py:118]: Why do we need this ? I'd like to avoid pattern like an other object calling an index.is_end_to_end_index such thing add debt. An other object from another module should not call attributes of indexes. Also I dislike class variable. It could be a property of the class if there really no choice and then update the abstract mother class but I'd rather like to avoid this

File: pylate/losses/cached_contrastive.py [hunk 4]
```diff
@@ -259,76 +274,58 @@ def calculate_loss(self, reps, masks, with_backward: bool = False) -> Tensor:
...
-            labels = labels + rank * batch_size
-        losses: list[torch.Tensor] = []
         do_query_expansion = (
             self.model.do_query_expansion
             if hasattr(self.model, "do_query_expansion")
             else self.model.module.do_query_expansion
         )
+
+        N = len(embeddings_other)
+        docs_stacked = torch.stack(embeddings_other, dim=1)
```
Comment (raphaelsty) [inline pylate/losses/cached_contrastive.py:284]: For claude, double check we need a torch.stack here and if we could not achieve the same thing with a torch.cat somehow. High propability stack is the best option, just check :)

File: pylate/retrieve/xtr.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    ...     device="cpu",
+    ... )
+
+    >>> assert isinstance(results, list)
+    >>> assert len(results) == 1
+
+    """
+
+    def __init__(self, index: indexes.Base, verbose: bool = False) -> None:
+        if index.is_end_to_end_index:
```
Comment (raphaelsty) [inline pylate/retrieve/xtr.py]: I got it but there is a design issue here. Also you need a PhD in IR to understand the error. This is_end_to_end_index is my primary concern so far

File: pylate/retrieve/xtr.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+import numpy as np
+import torch
+from tqdm.auto import tqdm
+
+from .. import indexes
+from ..rank import RerankResult, score_xtr
+from ..utils import iter_batch
+
+
+class XTR:
```
Comment (raphaelsty) [inline pylate/retrieve/xtr.py]: The class is beautiful

File: pylate/scores/colbert.py [hunk 7]
```diff
@@ -203,3 +203,223 @@ def colbert_kd_scores(
 
     scores = scores.max(axis=-1).values.sum(axis=-1)
     return scores
+
+
+class ColBERTScores:
```
Comment (raphaelsty) [inline pylate/scores/colbert.py:208]: Please don't repeat the module name in the class name

```python
from pylate import scores

scores.ColBERT
```

is better than 

```python
from pylate import scores

scores.ColBERTScores # We get it that we are talking about scores, maybe we should add it multiple times like to make sure user understand

scores.ColBERTScoresThIsisTalkingAboutScoresYesScoresYouheardWellScores
```

I mean this is a big MR, so it's ok to joke a bit.

Here in the score module I expect a colbert.py file and an xtr.py file with a classs ColBERT and a class XTR. So please create two files from your additions.

File: pylate/scores/scores.py [hunk 8]
```diff
@@ -203,3 +203,223 @@ def colbert_kd_scores(
...
+                documents_mask[:, j] if documents_mask is not None else None,
+            )
+            for j in range(N)
+        ]
+        # Stack to (Q_query, D, N) then flatten to (Q_query, D*N) with
+        # query-major ordering (doc d's k-th slot at column d*N + k).
+        return torch.stack(per_group, dim=2).reshape(-1, D * N)
+
+
+# Adapted from PrimeQA (https://github.com/primeqa/primeqa branch:xtr)
```
Comment (raphaelsty) [inline pylate/scores/scores.py]: Put this in the docstring please, there is a references field which the doc parse, otherwise no one will read this

File: pylate/scores/scores.py [hunk 9]
```diff
@@ -203,3 +203,223 @@ def colbert_kd_scores(
...
+
+        if queries_mask is not None:
+            topk_scores_max = topk_scores_max * queries_mask.unsqueeze(-1)
+
+        scores_sum = topk_scores_max.sum(dim=1)  # (Qb, Db)
+        Z = topk_scores_max.gt(0).float().sum(dim=1).clamp_(min=1e-3)
+        return (scores_sum / Z).float()
+
+
+class XTRKDScores(XTRScores):
```
Comment (raphaelsty) [inline pylate/scores/scores.py]: KDXTR would be better naming I think so, put this in xtr.py within scores module.

**Review by raphaelsty — COMMENTED** [review]
Amazing MR, this is a very cool feature, I leaved some comments, tried to be funny, open for discussion ofc. I think overall the MR missed something which is telling peoples how to use this index and when to use it and with which model appart from the comments I made :)

File: pylate/indexes/__init__.py (hunk 2 above)
Comment (NohTow) [reply pylate/indexes/__init__.py]: I think it comes from @robro612 
no strong opinion if it's not needed

Comment (robro612) [reply pylate/indexes/__init__.py]: I think I exposed it because I wanted to generalize the type hints in the retriever class. I think that could be done hidden without exposing it in the api, or just omitting it.

File: pylate/indexes/plaid.py (hunk 3 above)
Comment (NohTow) [reply pylate/indexes/plaid.py:118]: We discussed this and we can still discuss it. I would rather have a proper attribute than typechecking in Python, which is brittle and already broke before.

File: pylate/retrieve/xtr.py (hunk 5 above)
Comment (NohTow) [reply pylate/retrieve/xtr.py]: let's call after to discuss it

File: pylate/retrieve/xtr.py (hunk 6 above)
Comment (NohTow) [reply pylate/retrieve/xtr.py]: Used to be much worse, thanks!

File: pylate/indexes/plaid.py (hunk 3 above)
Comment (robro612) [reply pylate/indexes/plaid.py:118]: @NohTow at one point suggested the index class structure could be split into E2E/DocumentIndex and TokenIndex to distinguish the WARP/PLAID from regular old token-level indices like ScaNN. Then this check becomes an `isinstance` check, which is kind of what it already is in the retriever class. 

Bottom line: do you consider a class attribute to be meaningfully less desirable than a class itself?

File: pylate/scores/scores.py (hunk 8 above)
Comment (raphaelsty) [reply pylate/scores/scores.py]: Here is an exemple of numpy-style docstring :) 
```
  References
  ----------
  - [NanoBEIR](https://huggingface.co/collections/zeta-alpha-ai/nanobeir-66e1a0a
  f21dfd93e620cd9f6)
```

File: pylate/indexes/__init__.py (hunk 2 above)
Comment (NohTow) [reply pylate/indexes/__init__.py]: Solved for both indexes and retrieve

File: pylate/indexes/plaid.py (hunk 3 above)
Comment (NohTow) [reply pylate/indexes/plaid.py:118]: chatted it through, we go for an abstract property and will add another one for is_xtr/is_colbert later on.
It's a bit verbose imho but honestly not a big deal + it's clearer/safer. we shouldn't have too many "retrieval type", else we'll need a big refactor anyways so let's ship

File: pylate/scores/colbert.py (hunk 7 above)
Comment (NohTow) [reply pylate/scores/colbert.py:208]: We discussed it and although this is a very fair comment, we have had scores.colbert_scores for a long time and refactoring all of this would be scope creep.
open an issue/another pr if needed

File: pylate/scores/scores.py (hunk 8 above)
Comment (NohTow) [reply pylate/scores/scores.py]: should be good now

File: pylate/scores/scores.py (hunk 9 above)
Comment (NohTow) [reply pylate/scores/scores.py]: splitted the scores, did not rename, I honestly do not mind

**Comment by raphaelsty** [comment]
LGTM you can merge

**Comment by NohTow** [comment]
Thanks @robro612 @arn4 @raphaelsty, good job!
